lib/auth/auth_with_roles.go

```diff
@@ -5946,44 +5930,100 @@ func (a *ServerWithRoles) ReplaceRemoteLocks(ctx context.Context, clusterName st
 // channel if one is encountered. Otherwise the event channel is closed when the stream ends.
 // The event channel is not closed on error to prevent race conditions in downstream select statements.
 func (a *ServerWithRoles) StreamSessionEvents(ctx context.Context, sessionID session.ID, startIndex int64) (chan apievents.AuditEvent, chan error) {
-	createErrorChannel := func(err error) (chan apievents.AuditEvent, chan error) {
+	// StreamSessionEvents can be called internally, and when that happens we
+	// don't want to emit an event or check permissions.
+	if err := a.localServerAction(); err == nil {
+		return a.alog.StreamSessionEvents(ctx, sessionID, startIndex)
+	}
+
+	if err := a.actionForKindSession(ctx, apidefaults.Namespace, sessionID); err != nil {
+		c, e := make(chan apievents.AuditEvent), make(chan error, 1)
+		e <- trace.Wrap(err)
+		return c, e
+	}
+
+	eventsCh, errCh := a.alog.StreamSessionEvents(ctx, sessionID, startIndex)
+	return a.streamSessionEventsWithAudit(ctx, sessionID, eventsCh, errCh)
+}
+
+// streamSessionEventWithAudit audit log the recording access and streams the
+// session recording events.
+func (a *ServerWithRoles) streamSessionEventsWithAudit(ctx context.Context, sessionID session.ID, rawEventsCh chan apievents.AuditEvent, rawErrCh chan error) (chan apievents.AuditEvent, chan error) {
+	// Peek session events to determine the session type for audit logging.
+	// Also peek the error channel to avoid blocking the function call if the
+	// stream has errors (e.g. when session is not found).
+	eventsCh, firstEventCh := peekChannel(rawEventsCh, 1)
+	errCh, firstErrCh := peekChannel(rawErrCh, 1)
+
+	var sessionKind types.SessionKind
+	select {
+	case <-ctx.Done():
+	case <-firstErrCh:
```

Comment:
Not sure why we need to peek if the error handling is ignored 
The comments says 
> Also peek the error channel to avoid blocking

but the implementation will resent the error to errCh so I'm but consufed how this help to avoid blocking.

Reply:
If the `StreamSessionEvents` call makes an error, we never receive an event on the regular events channel (only one error on the error channel). If we don't peek at the error channel, we will only be unblocked when the context is finished.

As for peeking, we only don't forward the error channel if the emit audit event fails. If we peek, the caller will only be sent a message on the events channel that will never be sent instead of a channel receiving the error.

To clarify, we ignored the error because the audit log does not specify whether the session recording access was successful.

Flow:
```mermaid
sequenceDiagram
    participant C as Caller
    participant S as Server
    participant A as Audit log
    
    C->>S: StreamSessionEvents
    S->>A: StreamSessionEvents
    A->>S: (chan AuditEvent, chan error)
    S->>S: Peek error channel
    alt Receives event (on channel)
        S->>S: Determine session type
    else Receives error (on channel)
        S->>S: Does nothing
    else Context done
        S->>S: Does nothing
    end
    
    S->>S: EmitAuditEvent
    alt Emit fails
        S->>C: return new error channel<br/>and empty events channel
    else
        S->>C: peeked channels
    end
```